### PR TITLE
Support for changing database in model configuration

### LINKED ISSUE #110: Changing the database via {{ config(database='OTHER_DB') }} doesn't work
My default database name is `EDW_BDM` but I'm trying to create this one view in a different database `FI_DM_EBI`.

sqlserver.sql:
```
{{ config(database='FI_DM_EBI') }}
SELECT
  'a' AS col1,
  'b' AS col2
```

`dbt run` prints `START view model FI_DM_EBI.dbo.sqlserver` and dbt docs show below database structure:

image

However, the view still gets created in `EDW_BDM`. Looking at debug output, the connection string used is the default one that has the database `EDW_BDM`.

The same thing happens when changing the database name in `dbt_project.yml`:
```
models:
  hopedw:
    example:
      database: FI_DM_EBI
```

Please add support for changing the database.

### PR DESCRIPTION
**Pull request by semcha**
Fixes #110.

### CODE REVIEW

**Comment by semcha** [comment]
@swanderz Why circleci run only unit test?

**Comment by swanderz** [comment]
> @swanderz Why circleci run only unit test?

not sure it was saying "unauthorized". I was able to manually re-run the checks and they are passing now. but wanna take a look at this list and see if some might apply to you? You have a CircleCI account right?
https://support.circleci.com/hc/en-us/articles/360050273651-Builds-Unauthorized-due-to-contexts

File: dbt/include/sqlserver/macros/adapters.sql
Comment (swanderz) [inline dbt/include/sqlserver/macros/adapters.sql:27]: any reason you changed this?

```suggestion
      end as table_type
```

File: dbt/include/sqlserver/macros/adapters.sql
Comment (swanderz) [inline dbt/include/sqlserver/macros/adapters.sql:131]: @semcha was this prompted by an error?  I'm not sure why dbt would ever rename tables across databases? I'm pretty sure this macro is only used by dbt for renaming temp tables to their final names

File: dbt/include/sqlserver/macros/adapters.sql
Comment (swanderz) [inline dbt/include/sqlserver/macros/adapters.sql:124]: @semcha was the reason for this to make the macro run in a single transaction?

**Review by swanderz — COMMENTED** [review]
The CircleCI tests are passing for Azure SQL, but I'm think I'm going to test this out over the next few days.

File: dbt/include/sqlserver/macros/adapters.sql
Comment (semcha) [reply dbt/include/sqlserver/macros/adapters.sql:27]: PostgreSQL adapter uses field name "type":
https://github.com/fishtown-analytics/dbt/blob/develop/plugins/postgres/dbt/include/postgres/macros/adapters.sql#L63

File: dbt/include/sqlserver/macros/adapters.sql
Comment (semcha) [reply dbt/include/sqlserver/macros/adapters.sql:131]: I took this code from https://github.com/dbt-msft/dbt-sqlserver/pull/111 , but I think it's not necessary if this macros only for dbt use only

File: dbt/include/sqlserver/macros/adapters.sql
Comment (semcha) [reply dbt/include/sqlserver/macros/adapters.sql:124]: What is your suggestion? :)

Comment (swanderz) [reply dbt/include/sqlserver/macros/adapters.sql:124]: I think my suggestion would be:
```sql
{% macro sqlserver__create_view_as(relation, sql) -%}
  use [{{ relation.database }}];
  create view {{ relation.include(database=false) }} as
    {{ sql }}
{% endmacro %}
```

File: dbt/include/sqlserver/macros/adapters.sql
Comment (swanderz) [reply dbt/include/sqlserver/macros/adapters.sql:131]: @mikaelene can you see this being valuable? My thinking is that this macro is only used for changing a table called `example__dbt_tmp` to `example` at the end of the materialization.

File: dbt/include/sqlserver/macros/adapters.sql
Comment (semcha) [reply dbt/include/sqlserver/macros/adapters.sql:124]: @swanderz That way you get an error `'CREATE VIEW' must be the first statement in a query batch.`

**Review by swanderz — COMMENTED** [review]
@semcha have you tried running a dbt project after installing the adapter from this branch? I'd be interested if it works as we hope.

**Comment by semcha** [comment]
@swanderz Yep, it's currently running on my production server

**Comment by swanderz** [comment]
ok cool. so you're enabled in the short-term then. It's on my plate to see if this plays nicely with dbt-synapse. I hope to have time this week to do so. Thanks for all your work on this so far @semcha! this is the number one requested feature

**Review by mikaelene — APPROVED** [review]
Hi, This is all ok with me! @swanderz , do you have questions if this will work with Synapse to handle before merging?

**Comment by swanderz** [comment]
> Hi, This is all ok with me! @swanderz , do you have questions if this will work with Synapse to handle before merging?

yeah this is the last piece. I should have time to check in the next few days

**Comment by panasenco** [comment]
Thank you @semcha for this PR! I totally dropped the ball on this. I just experienced a `No item can be found` error that was fixed by switching to your branch, so your code is more reliable than mine. Great job, hope this gets merged soon!

**Comment by panasenco** [comment]
Hi, could someone go ahead and merge this? :)

**Comment by swanderz** [comment]
Hey @panasenco we definitely want to merge this, but I need to test these changes with dbt-synapse which I haven't had the time to do yet. I promise to do it this week sometime.

In the meantime, you can use this branch by installing this branch from GitHub or by cloning it locally and calling `pip install -e .` from the repo directory after checking out this branch.

**Comment by panasenco** [comment]
Yeah, I've been doing that, so no worries. :)

File: dbt/include/sqlserver/macros/adapters.sql
Comment (swanderz) [inline dbt/include/sqlserver/macros/adapters.sql:46]:
```suggestion
    USE [{{ relation.database }}];
```

File: dbt/include/sqlserver/macros/adapters.sql
Comment (swanderz) [inline dbt/include/sqlserver/macros/adapters.sql:87]:
```suggestion
  USE [{{ relation.database }}];
```

File: dbt/include/sqlserver/macros/adapters.sql
Comment (swanderz) [inline dbt/include/sqlserver/macros/adapters.sql:112]:
```suggestion
    USE [{{ relation.database }}];
```

File: CHANGELOG.md
Comment (swanderz) [inline CHANGELOG.md:11]: @mikaelene this ok by you?

File: dbt/include/sqlserver/macros/adapters.sql
Comment (swanderz) [inline dbt/include/sqlserver/macros/adapters.sql:37]:
```suggestion
    USE {{ database }};
```

File: dbt/include/sqlserver/macros/adapters.sql
Comment (swanderz) [inline dbt/include/sqlserver/macros/adapters.sql:119]:
```suggestion
    USE [{{ to_relation.database }}];
```

File: dbt/include/sqlserver/macros/adapters.sql
Comment (swanderz) [inline dbt/include/sqlserver/macros/adapters.sql:155]:
```suggestion
   USE [{{ relation.database }}];
```

**Review by swanderz — APPROVED** [review]
@semcha, I inititally made https://github.com/dbt-msft/dbt-synapse/pull/49 to prove to you that this PR, won't work only to realize that only a few changes would be necessary!

@mikaelene, this PR is ready to merge, provided you're okay with releasing a `v0.19.1` instead of a `v0.19.0.3`!

**Comment by visch** [comment]
Just ran into this myself, wasted a bunch of time on it. I should have checked this PR first. Thanks for the work here!

**Comment by swanderz** [comment]
I'll merge later today. @mikaelene will have to publish a release to PyPi though.

**Comment by visch** [comment]
Smoke test.
My problem may be attached to #135 , but after updating the macros in my project manually I noticed some curious behavior.

One of my source tables isn't populated, the model looks like this

```
with source as (

        select
        *
        from {{ source('raw', 'nodata') }}    

), final AS (
	select 
	*
	from source
)
select * from final
```

DBT reports this as running successfully after updating the macro. Even though my raw data doesn't exist. Looking at the dbt log and running the generated sql manually the command fails with something like this

```
Msg 208, Level 16, State 1, Procedure sourcemodel__dbt_tmp_temp_view, Line 6 [Batch Start Line 0]
Invalid object name 'raw.nodata'.
Msg 208, Level 16, State 1, Line 33
Invalid object name 'sourcemodel__dbt_tmp_temp_view'.

Completion time: 2021-05-25T12:20:51.9277966-06:00
```

**Comment by visch** [comment]
> Smoke test.
> My problem may be attached to #135 , but after updating the macros in my project manually I noticed some curious behavior.
> 
> One of my source tables isn't populated (I'd expect my staging model to fail) , the model looks like this
> 
> ```
> with source as (
> 
>         select
>         *
>         from {{ source('raw', 'nodata') }}    
> 
> ), final AS (
> 	select 
> 	*
> 	from source
> )
> select * from final
> ```
> 
> DBT reports this as running successfully after updating the macro. Even though my raw data doesn't exist. Looking at the dbt log and running the generated sql manually the command fails with something like this
> 
> ```
> Msg 208, Level 16, State 1, Procedure sourcemodel__dbt_tmp_temp_view, Line 6 [Batch Start Line 0]
> Invalid object name 'raw.nodata'.
> Msg 208, Level 16, State 1, Line 33
> Invalid object name 'sourcemodel__dbt_tmp_temp_view'.
> 
> Completion time: 2021-05-25T12:20:51.9277966-06:00
> ```

Just to be clear, this pull request makes the code better than it was before so I'd say go ahead! Everything works on my end, it's just an error caused by EXEC

**Comment by SportsTribution** [comment]
I'm running into problems that are similar to visch's.
Arbitrary syntax errors in models, like writing SLECT instead of SELECT, go unnotized. dbt does not report an error at all — but no model is created. This problem does not exist in dbt-sqlserver=0.19.0.1 and must have been introduced later.
I'm writing this here, because I think the problem might be related to aforementioned code changes introducing multiple sql commands without a GO in between. 
https://github.com/flyway/flyway/issues/718 
> It seems that if you execute multiple sql commands in the same jdbc statement (which happens if you don't separate each sql command with GO), and an error occurs in some DML command (for example INSERT) that is not the first one in the batch, then the MS jdbc driver does not throw any exception if you don't call getMoreResults() on the jdbc statement, until the result of the failing command is retrieved.

To be clear, I think it's Microsoft's fault, not yours! ;)
 
All the best
Hannes

**Comment by visch** [comment]
> I'm running into problems that are similar to visch's.
> Arbitrary syntax errors in models, like writing SLECT instead of SELECT, go unnotized. dbt does not report an error at all — but no model is created. This problem does not exist in dbt-sqlserver=0.19.0.1 and must have been introduced later.
> I'm writing this here, because I think the problem might be related to aforementioned code changes introducing multiple sql commands without a GO in between.
> [flyway/flyway#718](https://github.com/flyway/flyway/issues/718)
> 
> > It seems that if you execute multiple sql commands in the same jdbc statement (which happens if you don't separate each sql command with GO), and an error occurs in some DML command (for example INSERT) that is not the first one in the batch, then the MS jdbc driver does not throw any exception if you don't call getMoreResults() on the jdbc statement, until the result of the failing command is retrieved.
> 
> To be clear, I think it's Microsoft's fault, not yours! ;)
> 
> All the best
> Hannes

Yeah it's a problem. If you could put a test case together for them on https://github.com/dbt-msft/dbt-sqlserver/issues/135 I think it would help get to the bottom of it. I haven't had the time to prove myself :D

**Pull request merged**
